haskell/test/SMT.hs

```diff
@@ -0,0 +1,808 @@
+-- SPDX-FileCopyrightText: 2020 tqtezos
+-- SPDX-License-Identifier: MIT
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module SMT
+  ( stablecoinSMTSpec
+  ) where
+
+import Data.Maybe (fromJust)
+import qualified Data.Map as Map
+import Data.Typeable (cast)
+import Fmt
+
+import Test.Hspec (Spec, describe, it, expectationFailure)
+import Test.QuickCheck.Random (QCGen(..))
+import Test.QuickCheck (isSuccess, stdArgs, replay, usedSeed, quickCheckResult, quickCheckWithResult)
+import Test.Tasty.QuickCheck
+  ( Arbitrary, Gen, Property, arbitrary, withMaxSuccess, choose, arbitraryBoundedEnum, elements
+  , sublistOf, shuffle, vectorOf)
+import System.Random.SplitMix
+import qualified Text.Show
+import qualified Text.Read
+
+import Tezos.Core (unsafeMkMutez)
+
+import Lorentz (NiceParameter, arg, toVal, fromVal, def)
+import Lorentz.Address
+import Lorentz.Contracts.Stablecoin
+import qualified Lorentz.Contracts.Spec.FA2Interface as FA2
+import Lorentz.Contracts.Test.FA2
+import Lorentz.Contracts.Test.Common
+import Michelson.Test.Integrational
+import Michelson.Test.Dummy
+import Michelson.Interpret
+import Michelson.Typed (untypeValue)
+import Michelson.Text
+import Michelson.TypeCheck
+import qualified Michelson.Typed as T
+import qualified Michelson.Untyped as U
+import Util.Named
+
+-- Some implementation notes:
+--
+-- In this module we test the contract by implementing whole contract
+-- functionality in a Haskell, and comparing its behavior against the actual
+-- contract when run using a Michelson interpretr.
+--
+-- First we generate a random series of (sender address, contract parameter)
+-- pairs. Then we generated a random initial contract state.
+-- Then we apply each of these, both to the Haskell model, and on the
+-- michelson interpretation. After each iteration we check the contract state has not
+-- diverged. If it did, we indicate a test failure. If it did not, we extract the new
+-- contract state and initialize both models with it and apply next parameter, and so on.
+--
+-- For random input generation, we first assign random pool of addresses for
+-- each contract roles.  When we make a contract parameter, say transfer, we
+-- fetch a random sender address from the pool of token owners, and to/from
+-- address (for now we pick operator address and token owner address from the
+-- same pool). This is supposed to create inputs that have a little more
+-- chance of being valid input that has a chance to execute successfully.
+
+-- Right now safelist interaction is not tested, since it seems that doing it
+-- in a useful way require modeling a safelist contract as well.
+
+stablecoinSMTSpec
```

Comment:
Apparently here we test both FA2 and non-FA2 entrypoints (and behavior). But I intentionally created #42 and #43 as two separate issues because I want these tests to be separated. The primary reason is that one day we will move generic FA2 tests outside of this repo to use them in multiple projects.

Reply:
Yes, but it seemed to me that unlike Integrational tests, it would be less effective to have FA2 tests isolated here. Because we are running a sequence of contract inputs, and we need both FA2 calls along with the management calls in each of these sequences so that it will bring out bugs that only show up in some combination of these two.

I have separated FA2 tests into a separate commit now. Please check and let me know if we should further isolate it in some manner.

Reply:
See how it's done in [morley-ledgers-test](https://gitlab.com/morley-framework/morley/-/tree/master/code/morley-ledgers-test/src/Lorentz/Contracts/Test/SMT). We have SMT for `ApprovableLedger` and for `ManagedLedger`. The former tests only reduced functionality of FA1.2 tokens, but is still useful because it can detect bugs that can happen as result of calling only FA1.2 entrypoints. The latter can find more bugs, but can't be used for any FA1.2 contract.

Reply:
Well, actually, since FA2 is still quite unstable, I think current coupling is ok. One day FA2 will become stable and maybe we will implement another FA2 contract. Then we will decouple these tests. But for this PR I think decoupling is not required.